File: src/window.c
#include "window.h"
#include "camera.h"
#include "renderer.h"
#include <stdio.h>
#include <stdlib.h>

extern State state;

void window_mouse_callback(GLFWwindow *window, double x_pos, double y_pos) {
  Mouse *mouse = &state.window.mouse;

  (void)window;

  if (mouse->first_move_handled == false) {
    mouse->x = x_pos;
    mouse->y = y_pos;
    mouse->first_move_handled = true;
  }

  camera_mouse_move(mouse->x, x_pos, mouse->y, y_pos);

  mouse->x = x_pos;
  mouse->y = y_pos;
}

void window_scroll_callback(GLFWwindow *window, double x_offset,
                            double y_offset) {
  (void)window;
  (void)x_offset;

  camera_mouse_scroll((float)y_offset);
}

void window_key_callback(GLFWwindow *window, int key, int scancode, int action,
                         int mods) {
  (void)window;
  (void)scancode;
  (void)mods;

  if (key == GLFW_KEY_U && action == GLFW_PRESS) {
    renderer_toogle_wireframe();
  }
}

void window_init() {
  Window *window = &state.window;
  Keyboard *keyboard = &state.window.keyboard;
  Mouse *mouse = &state.window.mouse;

  window->width = 1600;
  window->height = 900;

  mouse->x = window->width / 2;
  mouse->y = window->height / 2;
  mouse->first_move_handled = false;

  keyboard->w_pressed = false;
  keyboard->a_pressed = false;
  keyboard->s_pressed = false;
  keyboard->d_pressed = false;

  if (!glfwInit()) {
    printf("Failed to initialize GLFW\n");
    exit(1);
  }

  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

  window->handle =
      glfwCreateWindow(window->width, window->height, "Game", NULL, NULL);

  if (!window->handle) {
    printf("Failed to create GLFW window\n");
    exit(1);
  }

  glfwMakeContextCurrent(window->handle);

  if (!gladLoadGLLoader((GLADloadproc)glfwGetProcAddress)) {
    printf("Failed to initialize GLAD\n");
    exit(1);
  }

  glfwSetInputMode(window->handle, GLFW_CURSOR, GLFW_CURSOR_DISABLED);

  glfwSetCursorPosCallback(window->handle, window_mouse_callback);
  glfwSetScrollCallback(window->handle, window_scroll_callback);
  glfwSetKeyCallback(window->handle, window_key_callback);

  glfwSetInputMode(window->handle, GLFW_LOCK_KEY_MODS, GLFW_TRUE);
}

void window_input() {
  glfwPollEvents();

  Keyboard *keyboard = &state.window.keyboard;

  if (glfwGetKey(state.window.handle, GLFW_KEY_ESCAPE) == GLFW_PRESS) {
    glfwSetWindowShouldClose(state.window.handle, true);
    return;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_W) == GLFW_PRESS) {
    keyboard->w_pressed = true;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_S) == GLFW_PRESS) {
    keyboard->s_pressed = true;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_A) == GLFW_PRESS) {
    keyboard->a_pressed = true;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_D) == GLFW_PRESS) {
    keyboard->d_pressed = true;
  }

  if (glfwGetKey(state.window.handle, GLFW_KEY_W) == GLFW_RELEASE) {
    keyboard->w_pressed = false;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_S) == GLFW_RELEASE) {
    keyboard->s_pressed = false;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_A) == GLFW_RELEASE) {
    keyboard->a_pressed = false;
  }
  if (glfwGetKey(state.window.handle, GLFW_KEY_D) == GLFW_RELEASE) {
    keyboard->d_pressed = false;
  }
}

File: src/main/java/voxelengine/core/Window.java
package voxelengine.core;

import org.joml.Vector2d;
import org.lwjgl.opengl.GL;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MAJOR;
import static org.lwjgl.glfw.GLFW.GLFW_CONTEXT_VERSION_MINOR;
import static org.lwjgl.glfw.GLFW.GLFW_CURSOR;
import static org.lwjgl.glfw.GLFW.GLFW_CURSOR_DISABLED;
import static org.lwjgl.glfw.GLFW.GLFW_CURSOR_NORMAL;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_A;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_D;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_ESCAPE;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_O;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_P;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_S;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_SPACE;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_U;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_W;
import static org.lwjgl.glfw.GLFW.GLFW_LOCK_KEY_MODS;
import static org.lwjgl.glfw.GLFW.GLFW_OPENGL_CORE_PROFILE;
import static org.lwjgl.glfw.GLFW.GLFW_OPENGL_PROFILE;
import static org.lwjgl.glfw.GLFW.GLFW_PRESS;
import static org.lwjgl.glfw.GLFW.GLFW_RELEASE;
import static org.lwjgl.glfw.GLFW.GLFW_TRUE;
import static org.lwjgl.glfw.GLFW.glfwCreateWindow;
import static org.lwjgl.glfw.GLFW.glfwDestroyWindow;
import static org.lwjgl.glfw.GLFW.glfwGetKey;
import static org.lwjgl.glfw.GLFW.glfwInit;
import static org.lwjgl.glfw.GLFW.glfwMakeContextCurrent;
import static org.lwjgl.glfw.GLFW.glfwPollEvents;
import static org.lwjgl.glfw.GLFW.glfwSetCursorPosCallback;
import static org.lwjgl.glfw.GLFW.glfwSetFramebufferSizeCallback;
import static org.lwjgl.glfw.GLFW.glfwSetInputMode;
import static org.lwjgl.glfw.GLFW.glfwSetKeyCallback;
import static org.lwjgl.glfw.GLFW.glfwSetWindowShouldClose;
import static org.lwjgl.glfw.GLFW.glfwSwapInterval;
import static org.lwjgl.glfw.GLFW.glfwTerminate;
import static org.lwjgl.glfw.GLFW.glfwWindowHint;
import static org.lwjgl.opengl.GL46.GL_DEPTH_TEST;
import static org.lwjgl.opengl.GL46.glEnable;
import static org.lwjgl.opengl.GL46.glViewport;
import static org.lwjgl.system.MemoryUtil.NULL;

public class Window {
    private long handle;
    private int width = 1600;
    private int height = 900;
    private final Vector2d mousePosition = new Vector2d((double) this.width / 2, (double) this.height / 2);
    private boolean firstMouseMoveHandled = false;
    private boolean cursorDisabled = true;

    public boolean wPressed = false;
    public boolean aPressed = false;
    public boolean sPressed = false;
    public boolean dPressed = false;
    public boolean spacePressed = false;

    public boolean isCursorDisabled() {
        return cursorDisabled;
    }

    public long getHandle() {
        return handle;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public void toggleCursor() {
        cursorDisabled = !cursorDisabled;
        glfwSetInputMode(this.handle, GLFW_CURSOR, cursorDisabled ? GLFW_CURSOR_DISABLED : GLFW_CURSOR_NORMAL);
    }

    public void init() {
        if (!glfwInit()) {
            throw new IllegalStateException("Unable to initialize GLFW");
        }

        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

        this.handle = glfwCreateWindow(this.width, this.height, "Voxel Engine", NULL, NULL);

        if (this.handle == NULL) {
            throw new IllegalStateException("Unable to create the GLFW window");
        }

        glfwMakeContextCurrent(this.handle);
        GL.createCapabilities();
        glfwSetInputMode(this.handle, GLFW_CURSOR, GLFW_CURSOR_DISABLED);
        glfwSetInputMode(this.handle, GLFW_LOCK_KEY_MODS, GLFW_TRUE);

        glfwSetCursorPosCallback(this.handle, (long window, double xpos, double ypos) -> {
            if (!this.firstMouseMoveHandled) {
                this.mousePosition.set(xpos, ypos);
                this.firstMouseMoveHandled = true;
            }

            // Only update camera if cursor is disabled (in game mode)
            if (cursorDisabled) {
                State.camera.handleMouseMove(this.mousePosition.x, xpos, this.mousePosition.y, ypos);
            }

            this.mousePosition.set(xpos, ypos);
        });

        glfwSetKeyCallback(this.handle, (window, key, scancode, action, mods) -> {
            if (key == GLFW_KEY_ESCAPE && action == GLFW_RELEASE) {
                glfwSetWindowShouldClose(window, true);
            }
            if (key == GLFW_KEY_U && action == GLFW_RELEASE) {
                State.renderer.toggleWireframe();
            }
            if (key == GLFW_KEY_O && action == GLFW_RELEASE) {
                State.physics.togglePhysics();
            }
            if (key == GLFW_KEY_P && action == GLFW_RELEASE) {
                toggleCursor();
            }
        });

        glfwSetFramebufferSizeCallback(this.handle, (long windowHandle, int newWidth, int newHeight) -> {
            this.width = newWidth;
            this.height = newHeight;
            glViewport(0, 0, this.width, this.height);
        });

        glfwSwapInterval(0);
        glEnable(GL_DEPTH_TEST);
    }

    public void input() {
        glfwPollEvents();

        // Only process movement inputs if cursor is disabled (in game mode)
        if (cursorDisabled) {
            if (glfwGetKey(this.handle, GLFW_KEY_W) == GLFW_PRESS) {
                this.wPressed = true;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_S) == GLFW_PRESS) {
                this.sPressed = true;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_A) == GLFW_PRESS) {
                this.aPressed = true;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_D) == GLFW_PRESS) {
                this.dPressed = true;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_SPACE) == GLFW_PRESS) {
                this.spacePressed = true;
            }

            if (glfwGetKey(this.handle, GLFW_KEY_W) == GLFW_RELEASE) {
                this.wPressed = false;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_S) == GLFW_RELEASE) {
                this.sPressed = false;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_A) == GLFW_RELEASE) {
                this.aPressed = false;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_D) == GLFW_RELEASE) {
                this.dPressed = false;
            }
            if (glfwGetKey(this.handle, GLFW_KEY_SPACE) == GLFW_RELEASE) {
                this.spacePressed = false;
            }
        } else {
            this.wPressed = false;
            this.sPressed = false;
            this.aPressed = false;
            this.dPressed = false;
            this.spacePressed = false;
        }
    }

    public void destroy() {
        glfwFreeCallbacks(this.handle);
        glfwDestroyWindow(this.handle);
        glfwTerminate();
    }
}
